tortoise/ninja_tortoise.go

```diff
@@ -120,27 +133,27 @@ func NewNinjaTortoise(layerSize int, blocks Mesh, hdist int, log log.Log) *Ninja
 		TSupport:   map[votingPattern]int{},
 		TPattern:   map[votingPattern]map[types.BlockID]struct{}{},
 
-		TVote:              map[votingPattern]map[types.BlockID]vec{},
-		TTally:             map[votingPattern]map[types.BlockID]vec{},
+		TVote:              map[votingPattern]map[BlockIDLayerTuple]vec{},
+		TTally:             map[votingPattern]map[BlockIDLayerTuple]vec{},
 		TComplete:          map[votingPattern]struct{}{},
-		TEffectiveToBlocks: map[votingPattern][]types.BlockID{},
+		TEffectiveToBlocks: map[votingPattern][]BlockIDLayerTuple{},
 		TPatSupport:        map[votingPattern]map[types.LayerID]votingPattern{},
 	}
 
 	return trtl
 }
 
 func (ni *NinjaTortoise) saveOpinion() error {
-	for bid, vec := range ni.TVote[ni.PBase] {
+	for b, vec := range ni.TVote[ni.PBase] {
 		valid := vec == Support
-		if err := ni.SaveContextualValidity(bid, valid); err != nil {
+		if err := ni.SaveContextualValidity(b.Id(), valid); err != nil {
 			return err
 		}
 
 		if !valid {
-			ni.With().Warning("block is contextually invalid", log.BlockId(bid.String()))
+			ni.With().Warning("block is contextually invalid", log.BlockId(b.BlockID.String()))
 		}
-		events.Publish(events.ValidBlock{Id: bid.String(), Valid: valid})
+		events.Publish(events.ValidBlock{Id: b.String(), Valid: valid})
```

Comment:
since you changed it above, for consistency let's change here as well.
Also, why do we have both B.Id() and b.BlockID under this for loop?

Reply:
look at blockLayerTuple implementation